sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1345 @@
+        Initialization of the linear transformation. Possible options are
+        'pca', 'identity' and a numpy array of shape (n_features_a,
+        n_features_b).
+
+        pca:
+            ``n_features_out`` many principal components of the inputs passed
+            to :meth:`fit` will be used to initialize the transformation.
+
+        identity:
+            If ``n_features_out`` is strictly smaller than the
+            dimensionality of the inputs passed to :meth:`fit`, the identity
+            matrix will be truncated to the first ``n_features_out`` rows.
+
+        numpy array:
+            n_features_b must match the dimensionality of the inputs passed to
+            :meth:`fit` and n_features_a must be less than or equal to that.
+            If ``n_features_out`` is not None, n_features_a must match it.
+
+    warm_start : bool, optional, (default=False)
+        If True and :meth:`fit` has been called before, the solution of the
+        previous call to :meth:`fit` is used as the initial linear
+        transformation (``n_features_out`` and ``init`` will be ignored).
+
+    max_impostors : int, optional (default=500000)
+        Maximum number of impostors to consider per iteration. In the worst
+        case this will allow ``max_impostors * n_neighbors`` constraints to be
+        active.
+
+    neighbors_params : dict, optional (default=None)
+        Parameters to pass to a :class:`neighbors.NearestNeighbors` instance -
+        apart from ``n_neighbors`` - that will be used to select the target
+        neighbors.
+
+    weight_push_loss : float, optional (default=0.5)
+        A float in (0, 1], weighting the push loss. This is parameter ``μ``
+        in the journal paper (See references below). In practice, the objective
+        function will be normalized so that the push loss has weight 1 and
+        hence the pull loss has weight ``(1 - μ)/μ``.
+
+    impostor_store : str ['auto'|'list'|'sparse'], optional
+        list :
+            Three lists will be used to store the indices of reference
+            samples, the indices of their impostors and the (squared)
+            distances between the (sample, impostor) pairs.
+
+        sparse :
+            A sparse indicator matrix will be used to store the (sample,
+            impostor) pairs. The (squared) distances to the impostors will be
+            computed twice (once to determine the impostors and once to be
+            stored), but this option tends to be faster than 'list' as the
+            size of the data set increases.
+
+        auto :
+            Will attempt to decide the most appropriate choice of data
+            structure based on the values passed to :meth:`fit`.
+
+    max_iter : int, optional (default=50)
+        Maximum number of iterations in the optimization.
+
+    tol : float, optional (default=1e-5)
+        Convergence tolerance for the optimization.
+
+    callback : callable, optional (default=None)
```

Comment:
we rarely support this. Is there a particular reason why LMNN deserves it?

Reply:
This was thought of as a way for the user to be able to inspect / evaluate / store intermediate solutions, as one can do when training (deep) neural networks.